graph/src/data/graphql/validation_hash.rs

```diff
@@ -0,0 +1,415 @@
+            .expect("q3 is syntactically valid")
+            .into_static();
+        let q4 = parse_query(Q4)
+            .expect("q4 is syntactically valid")
+            .into_static();
+
+        assert_ne!(validation_hash(&q1), validation_hash(&q2));
+        assert_ne!(validation_hash(&q3), validation_hash(&q4));
+    }
+
+    #[test]
+    fn sort_fragment_spreads() {
+        const Q1: &str =
+            "{ things { ...A ...B } } fragment A on ThingsA { a } fragment B on ThingsB { b }";
+        const Q2: &str =
+            "{ things { ...B ...A } } fragment A on ThingsA { a } fragment B on ThingsB { b }";
+        let q1 = parse_query(Q1)
+            .expect("q1 is syntactically valid")
+            .into_static();
+        let q2 = parse_query(Q2)
+            .expect("q2 is syntactically valid")
+            .into_static();
+
+        assert_eq!(validation_hash(&q1), validation_hash(&q2));
+    }
+
+    #[test]
+    fn sort_fragment_definitions() {
+        const Q1: &str =
+            "{ things { ...A ...B } } fragment B on ThingsB { b } fragment A on ThingsA { a } ";
+        const Q2: &str =
+            "{ things { ...A ...B } } fragment A on ThingsA { a } fragment B on ThingsB { b }";
+        let q1 = parse_query(Q1)
+            .expect("q1 is syntactically valid")
+            .into_static();
+        let q2 = parse_query(Q2)
+            .expect("q2 is syntactically valid")
+            .into_static();
+
+        assert_eq!(validation_hash(&q1), validation_hash(&q2));
+    }
+
+    #[test]
+    fn literals_and_operation_name() {
+        const Q1: &str = "query a { things(where: { stuff_gt: 20 }) { a c b d } }";
+        const Q2: &str = "query b { things(where: { stuff_gt: 30 }) { a c b d } }";
+        let q1 = parse_query(Q1)
+            .expect("q1 is syntactically valid")
+            .into_static();
+        let q2 = parse_query(Q2)
+            .expect("q2 is syntactically valid")
+            .into_static();
+
+        assert_eq!(validation_hash(&q1), validation_hash(&q2));
+    }
+
+    #[test]
+    fn selection_set_into_query() {
+        const Q1: &str = "query a { things { a c b d } }";
+        const Q2: &str = "        { things { a c b d } }";
+        let q1 = parse_query(Q1)
+            .expect("q1 is syntactically valid")
+            .into_static();
+        let q2 = parse_query(Q2)
+            .expect("q2 is syntactically valid")
+            .into_static();
+
+        assert_eq!(validation_hash(&q1), validation_hash(&q2));
+    }
+
+    #[test]
+    fn extra_field() {
+        const Q1: &str = "{ things { a c b d e } }";
+        const Q2: &str = "{ things { a c b d   } }";
+        let q1 = parse_query(Q1)
+            .expect("q1 is syntactically valid")
+            .into_static();
+        let q2 = parse_query(Q2)
+            .expect("q2 is syntactically valid")
+            .into_static();
+
+        assert_ne!(validation_hash(&q1), validation_hash(&q2));
+    }
+
```

Comment:
This is a test that should pass but fails:

```rust
#[test]
fn same_name_different_thing() {
    const Q1: &str = "{ a { b } c { d } }";
    const Q2: &str = "{ a { b c d } }";
    let q1 = parse_query(Q1)
        .expect("q1 is syntactically valid")
        .into_static();
    let q2 = parse_query(Q2)
        .expect("q2 is syntactically valid")
        .into_static();

    assert_ne!(validation_hash(&q1), validation_hash(&q2));
}
```

Fails with:
```
---- data::graphql::validation_hash::tests::same_name_different_thing stdout ----
thread 'data::graphql::validation_hash::tests::same_name_different_thing' panicked at 'assertion failed: `(left != right)`
  left: `14383803433003651353`,
 right: `14383803433003651353`', graph/src/data/graphql/validation_hash.rs:385:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The reason this fails is because the hash is not injective. I'd recommend reading EIP-712 for a primer on this concept: [https://eips.ethereum.org/EIPS/eip-712](https://eips.ethereum.org/EIPS/eip-712) but the short of it is that both of these queries hash as "a b c d" and make no distinction between types/nesting. We need to feed the rest of the syntax into the hash (like the "{" characters) so that these cases are disambiguated.

Reply:
I copied the algo from the `shape_hash` function

Reply:
Then both need to be fixed. cc @lutter